json/src/de.rs

```diff
@@ -404,35 +394,60 @@ impl<R: Read> DeserializerImpl<R> {
         };
 
         // Make sure a digit follows the exponent place.
-        let mut expi = match try!(self.next_char_or_null()) {
-            c @ b'0' ... b'9' => { (c as u64) - (b'0' as u64) }
+        let mut exp = match try!(self.next_char_or_null()) {
+            c @ b'0' ... b'9' => { (c - b'0') as i32 }
             _ => { return Err(self.error(ErrorCode::InvalidNumber)); }
         };
 
         while let c @ b'0' ... b'9' = try!(self.peek_or_null()) {
             self.eat_char();
+            let digit = (c - b'0') as i32;
+
+            if overflow!(exp * 10 + digit, i32::MAX) {
+                // Really big exponent, just ignore the rest.
+                while let b'0' ... b'9' = try!(self.peek_or_null()) {
+                    self.eat_char();
+                }
+                exp = i32::MAX;
+                break;
+            }
 
-            expi = try_or_invalid!(self, expi.checked_mul(10));
-            expi = try_or_invalid!(self, expi.checked_add((c as u64) - (b'0' as u64)));
+            exp = exp * 10 + digit;
         }
 
-        let expf = if expi <= i32::MAX as u64 {
-            10_f64.powi(expi as i32)
+        let final_exp = if pos_exp {
+            starting_exp.saturating_add(exp)
         } else {
-            return Err(self.peek_error(ErrorCode::InvalidNumber));
+            starting_exp.saturating_sub(exp)
         };
 
-        if pos_exp {
-            res *= expf;
-        } else {
-            res /= expf;
-        }
+        self.visit_f64_from_parts(pos, significand, final_exp, visitor)
+    }
 
-        if pos {
-            visitor.visit_f64(res)
-        } else {
-            visitor.visit_f64(-res)
-        }
+    fn visit_f64_from_parts<V>(&mut self,
+                               pos: bool,
+                               significand: u64,
+                               exponent: i32,
+                               mut visitor: V) -> Result<V::Value>
+        where V: de::Visitor,
+    {
+        let f = match POW10.get(exponent.abs() as usize) {
+            Some(pow) => {
+                if exponent >= 0 {
+                    significand as f64 * pow
+                } else {
+                    significand as f64 / pow
+                }
+            }
+            None => {
+                if exponent >= 0 && significand != 0 {
+                    f64::INFINITY
```

Comment:
This matches the current behavior (as of 0.7.4). I filed #112 to change the behavior.

```rust
assert_eq!(f64::INFINITY, serde_json::from_str("1e1000").unwrap());
```